azure/services/inboundnatrules/inboundnatrules.go

```diff
@@ -85,19 +90,23 @@ func (s *Service) Reconcile(ctx context.Context) error {
 		portsInUse[*rule.InboundNatRulePropertiesFormat.FrontendPort] = struct{}{} // Mark frontend port as in use
 	}
 
-	specs := s.Scope.InboundNatSpecs(portsInUse)
-	if len(specs) == 0 {
-		return nil
-	}
-
 	// We go through the list of InboundNatSpecs to reconcile each one, independently of the result of the previous one.
 	// If multiple errors occur, we return the most pressing one.
 	//  Order of precedence (highest -> lowest) is: error that is not an operationNotDoneError (i.e. error creating) -> operationNotDoneError (i.e. creating in progress) -> no error (i.e. created)
 	var result error
-	for _, natRule := range specs {
-		// If we are creating multiple inbound NAT rules, we could have a collision in finding an available frontend port since the newly created rule takes an available port, and we do not update portsInUse in the specs.
-		// It doesn't matter in this case since we only create one rule per machine, but for multiple rules, we could end up restarting the Reconcile function each time to get the updated available ports.
-		// TODO: We can update the available ports and recompute the specs each time, or alternatively, we could deterministically calculate the ports we plan on using to avoid collisions, i.e. rule #1 uses the first available port, rule #2 uses the second available port, etc.
+	for _, spec := range specs {
+		// Find an available SSH port for the rule.
+		sshFrontendPort, err := getAvailableSSHFrontendPort(portsInUse)
+		if err != nil {
+			return errors.Wrapf(err, "failed to find available SSH Frontend port for NAT Rule %s in load balancer %s", spec.ResourceName(), spec.OwnerResourceName())
+		}
+		natRule, ok := spec.(*InboundNatSpec)
+		if !ok {
+			result = errors.Errorf("%T is not of type InboundNatSpec", spec)
+		}
+		natRule.SSHFrontendPort = &sshFrontendPort
+		// Add the SSH frontend port to the list of ports in use
+		portsInUse[sshFrontendPort] = struct{}{}
```

Comment:
So we're now keeping an accounting of this during the runtime enumeration instead of in the spec. That makes sense as any individual NAT rule won't have authoritative knowledge of the ports in use at any given time T for TCP 2201-2220.

It does seem like there may be an edge case in our accounting, however. If we store `portsInUse` and re-use during each iteration in `specs`, then we may be unnecessarily storing a `sshFrontendPort`  if the below `s.CreateResource()` fails (which is what would be responsible for definitively "reserving" the `sshFrontendPort` on the NAT rule). Because we're not aborting out of the iteration in the failure case we would continue along and then `getAvailableSSHFrontendPort(portsInUse)` would grab the _next_ port in the allowable range, leaving "holes" in the `SSHFrontendPort` across our set of NAT rules.

Not sure if this is actually a problem, but tl;dr I wonder if we should put `portsInUse[sshFrontendPort] = struct{}{}` only after we verify that `err` from `s.CreateResource(ctx, natRule, serviceName)` is `nil`.

Reply:
Indeed this scenario _could_ happen. However, I don't think it's a problem for the following reasons:

- the find available port function is designed to be resilient to "holes" and finds the first available port in the range. This is because when machines are deleted, their NAT rule / port allocation is also deleted, which does create "holes". The next machine that gets created needs to be able to use those free ports. If we were to always increment we'd eventually run out of free ports. Each reconcile function does a List for that reason, so we can always get the latest accurate state of things.
- In practice, we're only creating one NAT rule per control plane VM. There are rarely more than 5 control plane VMs, we have lots of extra space in the range. This also means that we're not actually looping over a set of specs since the length of specs is always 0 or 1 (but still good to future proof).
- CreateResource() is asynchronous meaning we likely will not know if the rule was actually created successfully until a later reconcile loop. Given that, it is safer to assume that a port is taken as soon as we start an operation to create a rule with it.

Reply:
Thanks for the explanation, I'm convinced!